File: services/academic-service/src/test/java/com/visor/school/academicservice/integration/TeacherAssignmentIntegrationTest.java
package com.visor.school.academicservice.integration;

import com.visor.school.academicservice.config.TestConfig;
import com.visor.school.academicservice.model.ClassStatus;
import com.visor.school.academicservice.model.ClassType;
import com.visor.school.academicservice.model.EmploymentStatus;
import com.visor.school.academicservice.model.Teacher;
import com.visor.school.academicservice.model.Term;
import com.visor.school.academicservice.repository.TeacherRepository;
import com.visor.school.academicservice.service.ClassService;
import com.visor.school.academicservice.service.TeacherService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@ActiveProfiles("test")
@Import(TestConfig.class)
@Transactional
class TeacherAssignmentIntegrationTest {

    @Autowired
    private TeacherService teacherService;

    @Autowired
    private ClassService classService;

    @Autowired
    private TeacherRepository teacherRepository;

    @Test
    void shouldCompleteTeacherCreationAndAssignmentFlow() {
        UUID userId = UUID.randomUUID();
        Teacher teacher = teacherService.createTeacher(userId, List.of("Bachelor's Degree"), List.of("Mathematics", "Physics"), LocalDate.of(2020, 1, 1), "Science");

        assertNotNull(teacher);
        assertNotNull(teacher.getEmployeeId());
        assertEquals(EmploymentStatus.ACTIVE, teacher.getEmploymentStatus());

        com.visor.school.academicservice.model.Class classEntity = classService.createSubjectClass("Mathematics 101", "Mathematics", 10, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null);

        assertNotNull(classEntity);
        assertEquals(ClassType.SUBJECT, classEntity.getClassType());
        assertEquals(10, classEntity.getGradeLevel());

        var savedTeacher = teacherRepository.findById(teacher.getId());
        assertTrue(savedTeacher.isPresent());
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/integration/TeacherAssignmentIntegrationTest.kt
package com.visor.school.academicservice.integration

import com.visor.school.academicservice.config.TestConfig
import com.visor.school.academicservice.model.*
import com.visor.school.academicservice.repository.ClassRepository
import com.visor.school.academicservice.repository.TeacherAssignmentRepository
import com.visor.school.academicservice.repository.TeacherRepository
import com.visor.school.academicservice.service.ClassService
import com.visor.school.academicservice.service.TeacherService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID

/**
 * Integration test for teacher creation and assignment flow
 * Tests: Teacher creation → Class creation → Teacher assignment → Class teacher designation
 */
@SpringBootTest
@ActiveProfiles("test")
@Import(TestConfig::class)
@Transactional
class TeacherAssignmentIntegrationTest @Autowired constructor(
    private val teacherService: TeacherService,
    private val classService: ClassService,
    private val teacherRepository: TeacherRepository,
    private val classRepository: ClassRepository,
    private val teacherAssignmentRepository: TeacherAssignmentRepository
) {

    @Test
    fun `should complete teacher creation and assignment flow`() {
        // Given - Create teacher
        val userId = UUID.randomUUID()
        val teacher = teacherService.createTeacher(
            userId = userId,
            qualifications = listOf("Bachelor's Degree"),
            subjectSpecializations = listOf("Mathematics", "Physics"),
            hireDate = LocalDate.of(2020, 1, 1),
            department = "Science"
        )

        assertNotNull(teacher)
        assertNotNull(teacher.employeeId)
        assertEquals(EmploymentStatus.ACTIVE, teacher.employmentStatus)

        // When - Create subject class for grades 7-12
        val classEntity = classService.createSubjectClass(
            className = "Mathematics 101",
            subject = "Mathematics",
            gradeLevel = 10,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        assertNotNull(classEntity)
        assertEquals(ClassType.SUBJECT, classEntity.classType)
        assertEquals(10, classEntity.gradeLevel)

        // Then - Verify teacher can be assigned as class teacher
        // Note: In a full implementation, we'd create a TeacherAssignment first
        // For now, we verify the class exists and teacher exists
        val savedClass = classRepository.findById(classEntity.id!!)
        assertTrue(savedClass.isPresent)
        val savedTeacher = teacherRepository.findById(teacher.id!!)
        assertTrue(savedTeacher.isPresent)
    }

    @Test
    fun `should create teacher and assign to multiple classes`() {
        // Given - Create teacher
        val userId = UUID.randomUUID()
        val teacher = teacherService.createTeacher(
            userId = userId,
            qualifications = listOf("Master's Degree"),
            subjectSpecializations = listOf("English", "Literature"),
            hireDate = LocalDate.of(2020, 1, 1)
        )

        // When - Create multiple subject classes
        val class1 = classService.createSubjectClass(
            className = "English 101",
            subject = "English",
            gradeLevel = 9,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        val class2 = classService.createSubjectClass(
            className = "English 102",
            subject = "English",
            gradeLevel = 10,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        // Then - Verify both classes created
        assertNotNull(class1)
        assertNotNull(class2)
        assertEquals("English", class1.subject)
        assertEquals("English", class2.subject)
    }

    @Test
    fun `should validate teacher assignment constraints`() {
        // Given - Create teacher and class
        val userId = UUID.randomUUID()
        val teacher = teacherService.createTeacher(
            userId = userId,
            qualifications = emptyList(),
            subjectSpecializations = listOf("Mathematics"),
            hireDate = LocalDate.of(2020, 1, 1)
        )

        val classEntity = classService.createSubjectClass(
            className = "Mathematics 101",
            subject = "Mathematics",
            gradeLevel = 10,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )

        // When - Create teacher assignment
        val assignment = TeacherAssignment(
            teacherId = teacher.id!!,
            classId = classEntity.id!!,
            isClassTeacher = false
        )

        // Then - Verify assignment can be created
        // In a full implementation, this would be saved via a service method
        assertNotNull(assignment)
        assertEquals(teacher.id!!, assignment.teacherId)
        assertEquals(classEntity.id!!, assignment.classId)
    }
}

